common/src/main/java/com/microsoft/identity/common/internal/result/ResultFuture.java

```diff
@@ -28,21 +59,55 @@ public boolean isDone() {
     @Override
     public T get() throws InterruptedException {
         mCountDownLatch.await();
+
+        if (null != mException) {
+            throw new RuntimeException(mException);
+        }
+
         return mResult;
     }
 
     @Override
-    public T get(long l, TimeUnit timeUnit) throws InterruptedException, TimeoutException {
+    public T get(final long l, @NonNull final TimeUnit timeUnit) throws InterruptedException, TimeoutException {
         if (mCountDownLatch.await(l, timeUnit)) {
+            if (null != mException) {
+                throw new RuntimeException(mException);
+            }
+
             return mResult;
         } else {
-            throw new TimeoutException();
+            throw new TimeoutException(
+                    "Timed out waiting for: "
+                            + l // duration
+                            + timeUnit.name() // units
+            );
         }
+    }
+
+    public synchronized void setException(@NonNull final Exception exception) {
+        mException = exception;
+        mCountDownLatch.countDown();
 
+        for (final BiConsumer<T, Throwable> consumer : mConsumers) {
+            consumer.accept(mResult, exception);
+        }
     }
 
-    public void setResult(T result) {
+    public synchronized void setResult(@Nullable final T result) {
         mResult = result;
         mCountDownLatch.countDown();
+
+        for (final BiConsumer<T, Throwable> consumer : mConsumers) {
+            consumer.accept(result, mException);
+        }
+    }
+
+    public synchronized void whenComplete(@NonNull final BiConsumer<T, Throwable> consumerToAdd) {
+        if (isDone()) {
+            consumerToAdd.accept(mResult, mException);
+        }
+
+        // TODO Should we still add the consumer even if the ResultFuture isDone() == true?
```

Comment:
nah.  You can return.

Reply:
Great --  I'll yank that out

Reply:
40c7fbca336602ea74b7fa06af766c94fb39fc4b